File: pglite/src/pglite/release/preload/dataRequest.ts
import type { DataFileEntry, RuntimeModule } from "../core/runtimeTypes.js";

export class DataRequest {
	private static readonly requests = new Map<string, DataRequest>();
	private readonly start: number;
	private readonly end: number;
	private readonly audio: number;
	private readonly module: RuntimeModule;
	private readonly byteArray: Uint8Array;
	private name: string | null = null;

	constructor(
		module: RuntimeModule,
		byteArray: Uint8Array,
		entry: DataFileEntry,
	) {
		this.module = module;
		this.byteArray = byteArray;
		this.start = entry.start;
		this.end = entry.end;
		this.audio = entry.audio ?? 0;
	}

	open(name: string): void {
		this.name = name;
		DataRequest.requests.set(name, this);
		this.module.addRunDependency?.(`fp ${name}`);
	}

	onload(): void {
		const bytes = this.byteArray.subarray(this.start, this.end);
		this.finish(bytes);
	}

	private finish(bytes: Uint8Array): void {
		if (!this.name) {
			throw new Error("DataRequest has no open file name");
		}
		const createDataFile = this.module.FS_createDataFile;
		if (!createDataFile) {
			throw new Error("FS_createDataFile is not available on module");
		}
		createDataFile(this.name, null, bytes, true, true, true);
		this.module.removeRunDependency?.(`fp ${this.name}`);
		DataRequest.requests.delete(this.name);
		void this.audio;
	}
}

File: src/main/java/io/github/hidekatsu_izuno/pglite_jdbc/pglite/release/preload/dataRequest.java
package io.github.hidekatsu_izuno.pglite_jdbc.pglite.release.preload;

import io.github.hidekatsu_izuno.pglite_jdbc.pglite.release.core.runtimeTypes;

public class dataRequest {
    private static final java.util.Map<String, dataRequest> requests = new java.util.concurrent.ConcurrentHashMap<>();
    private final runtimeTypes.RuntimeModule module;
    private final byte[] byteArray;
    private final runtimeTypes.DataFileEntry entry;
    private String name;

    public dataRequest(runtimeTypes.RuntimeModule module, byte[] byteArray, runtimeTypes.DataFileEntry entry) {
        this.module = module;
        this.byteArray = byteArray;
        this.entry = entry;
    }

    public void open(String fileName) {
        this.name = fileName;
        requests.put(fileName, this);
        this.module.addRunDependency("fp " + fileName);
    }

    public void onload() {
        if (name == null) {
            throw new IllegalStateException("DataRequest has no open file name");
        }
        var size = Math.max(0, entry.end() - entry.start());
        var bytes = new byte[size];
        System.arraycopy(byteArray, entry.start(), bytes, 0, size);
        finish(bytes);
    }

    private void finish(byte[] bytes) {
        this.module.FS_createDataFile(name, null, bytes, true, true, true);
        this.module.removeRunDependency("fp " + name);
        requests.remove(name);
    }
}
